clustered/server/src/main/java/org/ehcache/clustered/server/store/ServerStore.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.server.store;
+
+import org.ehcache.clustered.common.store.Chain;
+import org.ehcache.clustered.common.store.Link;
+
+/**
+ * Defines Server Side data structure for storing
+ * cache data.
+ * This is conceptually a multi-map which has hash as the key
+ * and a {@link Chain} as the mapping
+ * Chain is list of {@link Link}s such that the same key has
+ * multiple mappings attached.
+ *
+ * Physical representation looks something like this :
+ *
+ *  hash -> |seqNum1|payLoad| - |seqNum2|payLoad| - |seqNum3|payLoad|
+ *  hash -> |seqNum1|payLoad| - |seqNum2|payLoad|
+ *
+ *  This only allows append & replace operations rather than
+ *  normal map operations.
+ *
+ */
+public interface ServerStore {
+
+  /**
+   * Returns the Chain associated with the provided hash.
+   * An empty Chain if no mapping exists.
+   *
+   * @param key hashcode of the key
+   * @return the {@link Chain} associated with the hash
+   */
+  Chain get(long key);
+
+  /**
+   * Appends the provided binary to Chain associated with key atomically.
+   * While appending, the payLoad is stored in {@link Link} associated with
+   * a sequence number.
+   *
+   * @param key to which the payLoad has to be appended
+   * @param payLoad to be appended
+   */
+  void append(long key, byte[] payLoad);
+
+  /**
+   * The Chain associated with key, previous to append is returned.
+   * The operation atomically appends payLoad and returns the previously associated Chain
+   * with the key.
+   *  Following block of instructions are atomically performed.
+   *  {
+   *    Chain prevChain = get(key);
+   *    append(key, payLoad);
+   *    return prevChain;
+   *  }
+   *
+   * @param key to which the payLoad has to be appended
+   * @param payLoad to be appended
+   * @return the Chain associated with the key before payLoad was appended
+   */
+  Chain getAndAppend(long key, byte[] payLoad);
+
+  /**
+   * Replaces the provided Chain with the equivalent Chain present as per
+   * provided sequence Numbers for the provided key.
+   * This operation is not guaranteed to succeed.
+   * The replace is successful iff the Chain associated with the key has
+   * a sub-sequence of links present with the provided list of sequence numbers.
```

Comment:
Can the sequence to be replaced be present anywhere in the stored chain, or can we be more specific than that?  What about multiple occurrences?